src/csv.rs

```diff
@@ -0,0 +1,136 @@
+use std::io::Write;
+use std::{
+    fs::OpenOptions,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use crate::{
+    cpu::CPU,
+    logger::{self, Interface, Logger},
+};
+
+pub struct CSVWriter {
+    log_size_cutoff: i32,
+    path: String,
+    enabled: bool,
+    logger: Logger,
+}
+
+trait Writer {
+    fn write(&mut self, column: Column);
+    fn init(&mut self, column: Column);
+}
+
+trait Writable {
+    fn to_csv(&mut self) -> String;
+}
+
+impl Writable for CPU {
+    fn to_csv(&mut self) -> String {
+        format!(
+            "{},{},{},{},{},{},{},{},{}\n",
+            SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap_or(Duration::new(0 as u64, 1 as u32))
+                .as_secs(),
+            self.name,
+            self.number,
+            self.max_freq,
+            self.min_freq,
+            self.cur_freq,
+            self.cur_temp,
+            self.cur_usage,
+            self.gov
+        )
+    }
+}
+
+pub struct Column {
+    cpus: Vec<CPU>,
+}
+
+impl Writer for CSVWriter {
+    /// Writes out all the cpu data from the daemon to the csv file
+    ///
+    /// This method gets called every `daemon.settings.delay` millis or every `daemon.settings.delay_battery` millis when on battery
+    ///
+    /// Each time this method gets called it creates a new row in the csv file. If the csv file
+    /// gets larger than `self.log_size_cutoff` MB it will cease logging.
+    ///
+    /// If an error occurs it will log the error to the daemon logger.
+    fn write(&mut self, column: Column) {
+        if !self.enabled {
```

Comment:
If something calls the writer function, is should always write. The check should be done before to avoid the function overhead of checking after call.

Reply:
I see your point. Make the csv writer more generic using ideas like Writable. I can see that being pretty cool.